Sources/BraveWallet/Crypto/Stores/SwapTokenStore.swift

```diff
@@ -502,7 +509,147 @@ public class SwapTokenStore: ObservableObject {
+    } else if let swapErrorResponse {
+      // check balance first because error can be caused by insufficient balance
+      if let sellTokenBalance = self.selectedFromTokenBalance,
+         let sellAmountValue = BDouble(self.sellAmount.normalizedDecimals),
+         sellTokenBalance < sellAmountValue {
+        self.state = .error(Strings.Wallet.insufficientBalance)
+        return
+      }
+      // check if jupiterQuote fails due to insufficient liquidity
+      if swapErrorResponse.isInsufficientLiquidity {
+        self.state = .error(Strings.Wallet.insufficientLiquidity)
+        return
+      }
+      self.state = .error(Strings.Wallet.unknownError)
+      self.clearAllAmount()
+    } else { // unknown error, ex failed parsing jupiter quote.
+      self.state = .error(Strings.Wallet.unknownError)
+      self.clearAllAmount()
+    }
+  }
+  
+  @MainActor private func handleSolPriceQuoteResponse(_ response: BraveWallet.JupiterQuote, base: SwapParamsBase) async {
+    guard let route = response.routes.first,
+          base == .perSellAsset // entering buy amount is disabled for Solana swap
+    else { return }
+    self.jupiterQuote = response
+    let formatter = WeiFormatter(decimalFormatStyle: .balance)
+    if let selectedToToken {
+      buyAmount = formatter.decimalString(for: "\(route.otherAmountThreshold)", radix: .decimal, decimals: Int(selectedToToken.decimals)) ?? ""
+    }
+    
+    // No exchange rate is returned by Jupiter API, so we estimate it from the quote.
+    if let selectedFromToken,
+       let newFromAmount = formatter.decimalString(for: "\(route.inAmount)", radix: .decimal, decimals: Int(selectedFromToken.decimals)),
+       let newFromAmountWrapped = BDouble(newFromAmount),
+       let selectedToToken,
+       let newToAmount = formatter.decimalString(for: "\(route.otherAmountThreshold)", radix: .decimal, decimals: Int(selectedToToken.decimals)),
+       let newToAmountWrapped = BDouble(newToAmount) {
+      let rate = newToAmountWrapped / newFromAmountWrapped
+      selectedFromTokenPrice = rate.decimalDescription
+    }
+    
+    await checkBalanceShowError(jupiterQuote: response)
+  }
+  
+  @MainActor private func checkBalanceShowError(jupiterQuote: BraveWallet.JupiterQuote) async {
+    guard let sellAmountValue = BDouble(sellAmount.normalizedDecimals),
+          let selectedFromTokenBalance else {
+      return
+    }
+    // Check if balance is insufficient
+    guard sellAmountValue <= selectedFromTokenBalance else {
+      state = .error(Strings.Wallet.insufficientBalance)
+      return
+    }
+    // Cannot currently estimate SOL gas balance to verify
+    self.state = .swap
+  }
+  
+  @MainActor private func createSolSwapTransaction() async -> Bool {
+    guard let jupiterQuote,
+          let route = jupiterQuote.routes.first,
+          let accountInfo = self.accountInfo,
+          let selectedToToken else {
+      return false
+    }
+    let coin = await walletService.selectedCoin()
```

Comment:
`coin` in this function has to be `.sol` dont need to get it from `walletService`

Reply:
Updated in [9ee57fc](https://github.com/brave/brave-ios/pull/6915/commits/9ee57fc395ce22ca2409ed5dd82509857868afe2)